tests: coredump: Disable optimizations to prevent unexpected failures

jonathonpenix opened the pull request:
debug.coredump.logging_backend is currently failing for most of the targets I have tried when building with clang. Depending on the target, func_3 invokes undefined behavior by dereferencing addr (which is NULL) which can lead to the compiler optimizing out significant portions of the code, resulting in unexpected/incorrect failures.

Here, clang seems to inline func_3 into main then marks the inlined implementation as unreachable (due to the UB) and removes it and everything after it in main. So, we fall through to whatever code lies past main, resulting in a test failure (timeout) from what I've seen. GCC seems to do similar things, but creates an invalid opcode instruction so the test still succeeds.

clang is correct in both optimizing this behavior out and leaving buggy code behind, so disable optimizations for func_3 to keep things under control and prevent the incorrect failures.

jonathonpenix commented:
Just to add a bit more context/info, I tested this with an LLVM 16 build as well as a recent-ish build of LLVM's mainline. The test would timeout for me for the following boards when building with both versions of LLVM:
- `qemu_x86`
- `qemu_cortex_a53` (using local changes to get this building correctly)
- `qemu_riscv32`/`qemu_riscv64` (using https://github.com/zephyrproject-rtos/zephyr/pull/68259)

The test seemed to work fine for the  `qemu_cortex_m3` board + clang before this fix as clang would not optimize out the `udf` instruction. Clang also doesn't seem to optimize out the `k_panic()` call, but just adding `__no_optimization` seemed less invasive than forcing clang to always take that path.

With this change to the test, the test passes for all boards mentioned above for me locally.

The command I was using is as follows (with minor variations for the board, etc.):

`west twister -W -p qemu_x86 -s tests/subsys/debug/coredump/debug.coredump.logging_backend -x=CONFIG_COMPILER_RT_RTLIB=y -x=CONFIG_LLVM_USE_LD=y -x=CONFIG_OUTPUT_DISASSEMBLY=y`

jonathonpenix commented:
The `riscv32_virtual`, `hifive_unleashed`, and `hifive_unmatched` boards apparently didn't like this change originally (they were timing out, please see the old CI logs for reference: https://github.com/zephyrproject-rtos/zephyr/actions/runs/8196145699/job/22415885346?pr=69939).

IIUC, these all seem to use Renode for simulation, so I think these boards match the comment about `Call k_panic() directly so Renode doesn't pause execution` and match what is discussed in https://github.com/zephyrproject-rtos/zephyr/pull/40434. As mentioned in the commit message above, GCC didn't seem to generate the null dereference previously (for RISC-V, it looks like it was generating an `ebreak` instruction) so it seems like these boards might have coincidentally worked previously?

To fix the CI issues, I added checks for these boards to have them take the `k_panic()` path. Please take another look and please let me know if there is another solution that would be prefereable!

dcpleung commented on tests/subsys/debug/coredump/src/main.c:
```diff
@@ -25,12 +25,18 @@ void k_sys_fatal_error_handler(unsigned int reason, const z_arch_esf_t *pEsf)
+/* Turn off optimizations to prevent the compiler from optimizing this away
+ * due to the null pointer dereference.
+ */
+__no_optimization void func_3(uint32_t *addr)
```
I don't think you need `__no_optimization` if it is going to trigger `k_panic()`.

jonathonpenix replied:
`__no_optimization` is still needed for boards that won't go through `k_panic()` when building with clang (`qemu_x86` and `qemu_riscv32/64`, for example--these are where I first bumped into this). The boards I added to the `#if` are just those that were negatively impacted by `__no_optimization` being added (I added them to fix the CI check for this PR).

As an alternative, I think I could drop the `__no_optimization` and add a change to always take the `k_panic()` route when using clang, but I'm not sure if that is better.

dcpleung replied:
In that case, please split this into two commits. One is to add the #ifdef for RISCV board, and the other one adding `__no_optimization`.

jonathonpenix replied:
I have now split this into two separate commits!

jonathonpenix committed:
tests: coredump: Change additional boards to directly call k_panic()

The riscv_virtual, hifive_unleashed, and hifive_unmatched boards all
timeout on this test if they actually perform the null dereference when
they are simulated. Whether the null dereference happens seems to depend
on the compiler used as well as whether optimizations are enabled though.

To make this more consistent, handle these boards in the same way as others
which use Renode simulation and have them directly call k_panic().

Signed-off-by: Jonathon Penix <jpenix@quicinc.com>

jonathonpenix committed:
tests: coredump: Disable optimizations to prevent unexpected failures

debug.coredump.logging_backend is currently failing for most of the targets
I have tried when building with clang. Depending on the target, func_3
invokes undefined behavior by dereferencing addr (which is NULL) which
can lead to the compiler optimizing out significant portions of the code,
resulting in unexpected/incorrect failures.

Here, clang seems to inline func_3 into main then marks the inlined
implementation as unreachable (due to the UB) and removes it and everything
after it in main. So, we fall through to whatever code lies past main,
resulting in a test failure (timeout) from what I've seen. GCC seems to do
similar things, but creates an invalid opcode instruction so the test still
succeeds.

clang is correct in both optimizing this behavior out and leaving
buggy code behind, so disable optimizations for func_3 to keep things under
control and prevent the incorrect failures.

Signed-off-by: Jonathon Penix <jpenix@quicinc.com>

jonathonpenix commented:
Gentle ping
